*** hash_page.c	Fri Jul  6 16:53:31 2001
--- hash_page.new	Tue Sep 11 14:00:43 2001
***************
*** 1609,1615 ****
  {
  	HASH_CURSOR *hcp;
  	db_pgno_t pgno;
- 	u_int32_t flags;
  	int gotmeta, ret;

  	hcp = (HASH_CURSOR *)dbc->internal;
--- 1609,1614 ----
***************
*** 1622,1632 ****
  		if ((ret = __ham_release_meta(dbc)) != 0)
  			return (ret);

! 	flags = 0;
! 	if (DB_NONBLOCK(dbc))
! 		LF_SET(DB_LOCK_NOWAIT);
!
! 	ret = __db_lget(dbc, flags, pgno, mode, 0, &hcp->lock);

  	hcp->lock_mode = mode;
  	return (ret);
--- 1621,1627 ----
  		if ((ret = __ham_release_meta(dbc)) != 0)
  			return (ret);

! 	ret = __db_lget(dbc, 0, pgno, mode, 0, &hcp->lock);

  	hcp->lock_mode = mode;
  	return (ret);
